### PR TITLE
Bug 1973565: Add create resource extension

### PR DESCRIPTION
**Pull request by rawagner**
This PR adds a new extension point which allows to use custom component when creating a resource either from OLM installed operator page or any other `~new` page

The following static plugins are currently overriding the create operand page via Page/Route extension (which wont work once/if they move to dynamic):
Ceph
Local storage
Container security

There is also DBaaS plugin in development which want to make use of this functionality.

## Usage ##
Dynamic plugin specifies k8s model for which it wants to replace the create operand page with custom component.
 
```json
{
  "type": "console.resource/create",
  "properties": {
    "model": {
      "group": "operator.tekton.dev",
      "version": "v1alpha1",
      "kind": "TektonConfig"
    },
    "component": { "$codeRef": "fooComp" }
  }
}
```

In theory, multiple plugins can contribute to the same model. In that case we only use the first contribution. This case will be handled better once we add priority prop to all extensions.

@spadgett @christianvogt @vojtechszocs

### CODE REVIEW

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-operand.ts
Comment (spadgett) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-operand.ts:6]: Do we want to make this a general "create resource" extension point that is used for both OLM operands and list pages (for example, when the Create button is clicked on Home -> Search).

Comment (vojtechszocs) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-operand.ts:6]: We should strive for a good balance between the number of extension points vs. their generic-ness (more generic means more properties and thus more complexity). Extension properties wise, if the overlap/intersection between OLM operands and list pages is relatively high, +1 to have just one extension point.

Comment (christianvogt) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-operand.ts:6]: I think we can reuse this same extension across all resources for their `/~new` page. 
How would a new operand differ to any other new resource?

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-resource.ts
Comment (TheRealJon) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-resource.ts:20]: not new, but `appName` is kind of ambiguous. If we aren't already using "appName" elsewhere in the dynamic plugin API, now would be a good time to change it. It will be hard/impossible to change later.  IMO, it would clearer to call this csvName, operatorName, or something along those lines.

```suggestion
          /**
           * Name of the OLM Operator ClusterServiceVersion that manages this resource.
           * Defined if resource creation is initiated from the Operator details page.
           */
           csvName?: string;
```

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-resource.ts
Comment (TheRealJon) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-resource.ts:28]: nit, extra newline

**Review by TheRealJon — COMMENTED** [review]
I like that we generalized this to all resource creation. Just one concern about the match parameter name.

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-resource.ts
Comment (rawagner) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-resource.ts]: Thanks @TheRealJon ! I also didnt like the current param name `appName` instead of more obvious like `csvName` but I thought there is a good reason for it - good that you want to change it too.

The latest update has is using `csvName`

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-resource.ts
Comment (rawagner) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-resource.ts:28]: This follows style of all other extension declarations. We always add the newline after `// Type guards` comment

**Review by spadgett — APPROVED** [review]
/lgtm
/hold 

It would be good for @vojtechszocs or @christianvogt to take a look as well.

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-resource.ts
Comment (christianvogt) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-resource.ts:20]: I don't think this param should belong in such a generic extension.

**Review by christianvogt — COMMENTED** [review]
You've created an extension which must return a complete page content.
What if the extension is only meant to render the creation form. We can implement the resource creation page along with a header which uses this extension to render the creation component.
The operand creation page would use the same extension to render the creation component below its page header.


In doing it this way, the extension doesn't need to care if it

**Comment by rawagner** [comment]
> What if the extension is only meant to render the creation form. We can implement the resource creation page along with a header which uses this extension to render the creation component.

Yes, maybe, but what if they want to have a custom header ? I think its better to not hardcode the header component and leave it up to the consumer to use what they want.

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-resource.ts
Comment (rawagner) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-resource.ts:20]: `csvName` is currently used in the OLM form to get the actual `csv` resource and load some additional data from it. Im not sure if the parameter is really needed for plugins.

@TheRealJon @spadgett WDYT ?

Comment (christianvogt) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-resource.ts:20]: Having the contributor need to render a different header to remain consistent will be problematic. Unless there's a component that can be reused and it can use the route info from a hook instead of relying on it being passed in as a prop.

This is why I suggested we render the headers and then use the extension to render the form area only.

Comment (christianvogt) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-resource.ts:20]: @rawagner can you render the operands breadcrumbs in OLM before rendering the extension component? This would make it so that contributors don't need to worry about the rendering of breadcrumbs.

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-resource.ts
Comment (vojtechszocs) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-resource.ts]: Yes, now is a good time to change things.

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-resource.ts
Comment (spadgett) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-resource.ts:20]: `csvName` might not be specific enough if this is a cluster-scoped resource since `ns` would be undefined.

I think in practice this isn't needed. Presumably the same operator that contributed the custom resource is customizing the creation page using a plugin.

Comment (TheRealJon) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-resource.ts:20]: @rawagner @spadgett @christianvogt I think we probably won't even need the CSV for a custom Operand creation form. The only thing we use it for in the current creation form is to get OLM descriptors and display operator details in the sidebar. Descriptors become obsolete in the context of a custom creation form, and an operator author can provide whatever Operator info they want on their custom creation form.

It also makes sense to me that the user should only need to implement the main content of the page, excluding the header and breadcrumbs. Maybe we could implement extensions just for overriding page headers or breadcrumbs if it becomes an issue?

Comment (christianvogt) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/create-resource.ts:20]: It seems we want to surface the same form in different contexts (operand and details page, maybe even the catalog in the future). Definitely the extensions shouldn't need to worry about the breadcrumbs.

**Comment by rawagner** [comment]
so the OLM now renders breadcrumbs and the rest (header/body) is done by dynamic plugin as agreed with @christianvogt . 
The `csvName` param was removed thanks to this.

**Comment by rawagner** [comment]
@TheRealJon @christianvogt @spadgett please take a look again :)

**Comment by christianvogt** [comment]
@rawagner i see some components with custom breadcrumbs because they must be overriding the entire create page still?
Are you planning on migrating those pages to the new extension in the future as to not increase the size of this pr?

**Comment by rawagner** [comment]
> @rawagner i see some components with custom breadcrumbs because they must be overriding the entire create page still?
> Are you planning on migrating those pages to the new extension in the future as to not increase the size of this pr?

yes, the components which Im changing still use the old way of overriding the whole route so they use their own breadcrumbs.
I dont plan on migrating those to new extension as I expect this to be done by the package owner once they move do dynamic plugins

File: frontend/packages/operator-lifecycle-manager/src/components/operand/create-operand.tsx
Comment (TheRealJon) [inline frontend/packages/operator-lifecycle-manager/src/components/operand/create-operand.tsx:161]: Declaring a component inside of another component is an antipattern. It might be better to just load the extensions here separately.

File: frontend/packages/operator-lifecycle-manager/src/components/operand/create-operand.tsx
Comment (TheRealJon) [inline frontend/packages/operator-lifecycle-manager/src/components/operand/create-operand.tsx:187]: Passing a component as a prop is also an antipattern. We should just load the extensions in this component and conditionally render here.

File: frontend/public/components/create-resource.tsx
Comment (TheRealJon) [inline frontend/public/components/create-resource.tsx:38]: We can factor out this logic into a hook that returns a single create resource extension given the `match.params.plural` arg. That way it can be reused here and in CreateOperandPage.

**Review by TheRealJon — COMMENTED** [review]
See comments.

File: frontend/packages/operator-lifecycle-manager/src/components/operand/create-operand.tsx
Comment (TheRealJon) [inline frontend/packages/operator-lifecycle-manager/src/components/operand/create-operand.tsx:152]: If we don't use `csv.spec.displayName` in the breadcrumbs, we could move this into CreateOperand, along with the StatusBox wrapper. I think it would make more sense this way because if there's a matching create resource extension, we are needlessly loading the CSV.

File: frontend/packages/operator-lifecycle-manager/src/components/operand/create-operand.tsx
Comment (TheRealJon) [inline frontend/packages/operator-lifecycle-manager/src/components/operand/create-operand.tsx:53]: Need to make sure model is defined before we use it to fetch CRD:
```suggestion
  const [crd] = useK8sWatchResource<CustomResourceDefinitionKind>(
    model
      ? {
          kind: CustomResourceDefinitionModel.kind,
          isList: false,
          name: nameForModel(model),
        }
      : null,
  );
```

**Comment by christianvogt** [comment]
DBaaS needs this change in 4.8 to let them override their operands page.

File: frontend/packages/operator-lifecycle-manager/src/components/operand/create-operand.tsx
Comment (TheRealJon) [inline frontend/packages/operator-lifecycle-manager/src/components/operand/create-operand.tsx:131]: No need to conditionally render on `loaded`, StatusBox takes care of that. I think the React fragment wrapper can be removed too.

File: frontend/public/components/utils/create-resource-hook.ts
Comment (TheRealJon) [inline frontend/public/components/utils/create-resource-hook.ts:16]: This should probably be in `frontend/packages/console-shared/src/hooks`.

**Review by TheRealJon — COMMENTED** [review]
Thanks for the updates! Just a couple more small comments and this should be good to go.

**Comment by rawagner** [comment]
@TheRealJon updated!

**Review by TheRealJon — APPROVED** [review]
/lgtm

**Pull request merged**
